### ISSUE TITLE
canActivate callbacks with router.add() POJO?

### ISSUE THREAD

**Issue by nicholaswyoung** [router5/router5#39]
```javascript
const routes = [
  {
    name: 'page',
    path: '/*',
    canActivate: (to, from, done) => {},
    onActivate: (dispatch) => (params) {},
  }
];

router.add(routes);
```

In the case above, where `routes` is a POJO that is then later added to the router, `canActivate` is never fired.

Is this an acceptable way to define routes, or should I be using `RouteNode` objects?

**Comment by troch** [comment]
I plan to improve / deprecate some API with 2.0.0 around adding routes, canActivate, etc...
Currently the only way to use add canActivate is with:
- `router.addNode(name, path, canActivate)
- `router.canActivate(name, canActivate)`

**Comment by nicholaswyoung** [comment]
Should I just ignore the POJO API for now? 

My main reason for investigating this API is universal rendering. I have several Redux action creators that return Promises. The hope is that I can use activation handlers to make sure all of the promises resolve before the component is rendered. 

Maybe this is the wrong approach, but at the very least, I now understand the Router5 API a little better. ;)

**Comment by troch** [comment]
I define my routes that exact same way (using an array of POJOs containing `canActivate` and / or `onActivate`). But for adding them, I do `routes.forEach(route => router.addNode(route.name, route.path, route.canActivate)`.

For `onActivate`, with Redux, and with the aim to resolve all promises before your new route ends up in your state, you just need a redux middleware invoking your onActivate handlers and wait for all of them to resolve before calling `next(action)` (action being `TRANSITION_SUCCESS`).

Basically the same example than [http://router5.github.io/docs/async-data.html#using-a-state-container-redux-](http://router5.github.io/docs/async-data.html#using-a-state-container-redux-) but with something similar to `Promise.all([/* your onActivate handler promises */]).then(() => next(action)))` (+ error handling).

**Comment by nicholaswyoung** [comment]
Would it be possible to fire the Redux `dispatch` call before the route activates? I see the ideal process looking something like this:

1. The incoming request is matched against known `RouteNode`s and their respective path. If a match is found, that matched `RouteNode` executes a callback, much like the existing `onActivate` handler.

2. Once the dispatch is finished, then Router5 continues.

3. The request is rendered, now with appropriate state.

**Comment by troch** [comment]
There is another action which is dispatched `TRANSITION_START`, before `TRANSITION_SUCCESS`. Not sure if this is what you are after, is that something you are trying to achieve on the server-side?

**Comment by nicholaswyoung** [comment]
Again, I may be going down the wrong path. I've tried several approaches to waiting until all of the Promise handlers have been resolved (see [redux-promise](https://github.com/acdlite/redux-promise/blob/master/src/index.js)) before rendering the server response. 

I figure that if I can wait for Promise resolution on the `TRANSITION_START` event, then maybe that will do the trick. As it stands, the Promises are resolving, but not in time for the state to be stringified and sent down along with the initial response.

**Comment by troch** [comment]
I have not yet pre-rendered on the server-side with redux, so not sure what I say is relevant. Is that the process you follow: `create your router` -> `create your store` -> `start the router` -> `renderString in router.start callback` -> `wrap it up`?

**Comment by troch** [comment]
Oh yeah I see the problem now.

**Comment by nicholaswyoung** [comment]
I'm actually creating the store inside the `router.start()` callback, so that I can inject Router5's `state` object as the router state (for client side syncing). But other than that, your pipeline example is exactly what I'm doing.

**Comment by nicholaswyoung** [comment]
See here: https://github.com/nicholaswyoung/hotwax/blob/redux/src/ui/renderer.js#L41-L73

**Comment by troch** [comment]
So the solution is: instead of using a redux middleware (after transition success), you can use a router5 middleware.

In it, you just do the same thing, and either call done or return a promise. Problem: `store` is not passed to middlewares. I have started to work on 2.0.0 (https://github.com/router5/router5/pull/40), but cannot fix middlewares now as it is a breaking change. 2.0.0 is quite small and will be out in one week - two weeks max.

In the meantime, there is a workaround. From within a middleware, you can access your store:
```javascript
const dataMiddleware = function(toState) {
    const { dispatch } = this.router._extraArgs[0];

    // return a Promise for delaying transition
};
```

**Comment by nicholaswyoung** [comment]
I just loaded in the middleware you suggested, and it looks like `extraArgs` is an empty array. I think I may have missed a step.

**Comment by troch** [comment]
I have tested it, works for me.

> I'm actually creating the store inside the router.start()

Maybe that is the issue ;)?

**Comment by nicholaswyoung** [comment]
Yeah, that was it. Thanks for pointing out the obvious bug. I guess that I've spent so much time working on this issue that I'm going blind to small, but important issues like that. 

I tried it out, but now I seem to be encountering an issue where `useMiddleware` blocks the whole process.

I'll investigate further, but for now, I think this issue is resolved. I just need to spend a bit more time in the docs. Thank you!

**Comment by troch** [comment]
OK, let me know how it goes.

**Comment by nicholaswyoung** [comment]
Will do. 

Also, I was wondering where I got the idea to create the store inside of the `router.start()` callback. It was in the [react-redux sample app](https://github.com/router5/examples/blob/master/apps/react-redux/main.js#L12).
